lib/active_model/serializer/reflection.rb

```diff
@@ -78,9 +86,15 @@ def value(serializer)
           if block_value != :nil
```

Comment:
in JSONAPI::Resources, the relationship object has a `foreign_key_on` option that may be `self`. 

[docs](https://github.com/cerebris/jsonapi-resources/blob/f4ab5683de49c2a506eea14096bd81a0ec85e94e/README.md#options)

> The relationship methods (`relationship`, `has_one`, and `has_many`) support the following options:
>  * `foreign_key` - the method on the resource used to fetch the related resource. Defaults to `<resource_name>_id` for has_one and `<resource_name>_ids` for has_many relationships.
> `to_one` relationships support the additional option:
>  * `foreign_key_on` - defaults to `:self`. To indicate that the foreign key is on the related resource specify `:related`.
> Examples:
> ```ruby
> class ExpenseEntryResource < JSONAPI::Resource
>   has_one :currency, class_name: 'Currency', foreign_key: 'currency_code'
> end
> ```


 see https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/relationship.rb for full code

```ruby
module JSONAPI
  class Relationship
    attr_reader :foreign_key

    def initialize(name, options = {})
      @foreign_key = options[:foreign_key] ? options[:foreign_key].to_sym : nil
    end

    def belongs_to?
      false
    end

    class ToOne < Relationship
      attr_reader :foreign_key_on

      def initialize(name, options = {})
        super
        @foreign_key ||= "#{name}_id".to_sym
        @foreign_key_on = options.fetch(:foreign_key_on, :self)
      end

      def belongs_to?
        foreign_key_on == :self
      end

    class ToMany < Relationship
      def initialize(name, options = {})
        super
        @foreign_key ||= "#{name.to_s.singularize}_ids".to_sym
      end
    end
  end
end
```